src/main/java/appeng/core/features/registries/grinder/GrinderRecipeManager.java

```diff
@@ -393,4 +399,123 @@ else if( this.item != other.item )
 		}
 
 	}
+
+	/**
+	 * Internal {@link IInscriberRecipeBuilder} implementation.
+	 * Needs to be adapted to represent a correct {@link IInscriberRecipe}
+	 */
+	private static final class Builder implements IGrinderRecipeBuilder
+	{
+
+		private ItemStack in;
+		private ItemStack out;
+
+		private float optionalChance;
+		private ItemStack optionalOutput;
+
+		private float optionalChance2;
+		private ItemStack optionalOutput2;
+
+		private int turns;
+
+		@Override
+		public IGrinderRecipeBuilder withInput( ItemStack input )
+		{
+			Preconditions.checkNotNull( input );
+
+			this.in = this.copy( input );
+
+			return this;
+		}
+
+		@Override
+		public IGrinderRecipeBuilder withOutput( ItemStack output )
+		{
+			Preconditions.checkNotNull( output );
+
+			this.out = this.copy( output );
+
+			return this;
+		}
+
+		@Override
+		public IGrinderRecipeBuilder withFirstOptional( ItemStack optional, float chance )
+		{
+			Preconditions.checkNotNull( optional );
+			Preconditions.checkArgument( chance >= 0 && chance <= 1.0 );
+
+			this.optionalOutput = this.copy( optional );
+			this.optionalChance = chance;
+
+			return this;
+		}
+
+		@Override
+		public IGrinderRecipeBuilder withSecondOptional( ItemStack optional, float chance )
+		{
+			Preconditions.checkNotNull( optional );
+			Preconditions.checkArgument( chance >= 0 && chance <= 1.0 );
+
+			this.optionalOutput2 = this.copy( optional );
+			this.optionalChance2 = chance;
+
+			return this;
+		}
+
+		@Override
+		public IGrinderRecipeBuilder withTurns( int turns )
+		{
+			Preconditions.checkArgument( turns > 0 );
+
+			this.turns = turns;
+
+			return this;
+		}
+
+		@Nonnull
+		@Override
+		public IGrinderRecipe build()
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') The Cyclomatic Complexity of this method "build" is 12 which is greater than 10 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)